### PR TITLE
fix(dot/network): fix receiving notifications messages

### LINKED ISSUE #1498: receive notifications messages on block-announce, grandpa, transactions substreams
## Task summary

- currently, it seems we are not receiving messages (apart from handshakes) on notifications substreams
- this task is to fix the underlying issue and receive notifications messages on block-announce, grandpa, transactions substreams

### PR DESCRIPTION
**Pull request by noot**
## Changes

- fix receiving notifications messages by writing handshake back on same stream it was received on, previously was opening a new substream on the same protocol
- add more retries to reconnecting to persistent peers
- remove sending handshake on connection logic, since the outbound stream should only be opened when we have a message to send the remote side

## Tests

```
go test ./dot/network -short
```

turn on network debug logs and:
```
make gossamer
./bin/gossamer --chain polkadot
```

should see `BlockAnnounce` messages being received, eg:
```
DBUG[04-12|15:13:54] received message on notifications sub-protocol pkg=network protocol=/dot/block-announces/1 message="BlockAnnounceMessage ParentHash=0xf54e9f260146d31ae06d497f8e4af0c2331508c2c9fccdf6035c4ab129d2b293 Number=4604640 StateRoot=0xda4c17beee3539ab91b2db1f0448c127b6de51e68908fb0d91df0051c0838881x ExtrinsicsRoot=0xc30c8275ad2dddfb63dabd4c09f3492c5c1be64a9e7a8e5e75c0ce3b2f698bfe Digest=[PreRuntimeDigest ConsensusEngineID=BABE Data=0x03cf000000536f13100000000012712e0e26ee15be11e594247bff7cbd42c9918754f06bb00c359a2779440832ff8c391a338f52e5fcca1dc5a9ed479963a54f844af05322682abbf1b302ae0a351bfafc483c54d27edfff86c663a953052c7811d6baa099e8fabe4b3ff2760d SealDigest ConsensusEngineID=BABE Data=0x4e08188d3c686f51693d2123e78cf051f87f861f89bacdd68004a5753831d36c7dfb427ff82cf47680df56e2964a267765973414640b67bd5a0a81d999828a80]" peer=12D3KooWJUAd5qeice2mQcwPFt7mEmwreP5SyTgpao3HPUqpdXvy caller=notifications.go:187
```

you will also see error logs such as this:
```
DBUG[04-12|15:13:49] failed to handle message from stream     pkg=network message="BlockAnnounceHandshake Roles=1 BestBlockNumber=184285 BestBlockHash=0x71442f581f4b8df2adbef617345844892cad76286c21684a85c6075fb8a580ea GenesisHash=0xa83ce41b713cfc4780263fbf1e396453fdd0bfa1106853f84f1675aacb6ea415" error="failed to validate handshake" caller=service.go:520
```
this is expected because the peer's genesis hash is not the polkadot genesis hash.

## Issues

- closes #1498

### CODE REVIEW

File: dot/network/service.go [hunk 1]
```diff
@@ -541,21 +509,18 @@ func (s *Service) readStream(stream libp2pnetwork.Stream, peer peer.ID, decoder
...
-			if err != nil {
-				logger.Trace("Failed to handle message from stream", "message", msg, "error", err)
-				_ = stream.Close()
-				return
-			}
-		}()
+		err = handler(stream, msg)
+		if err != nil {
+			logger.Debug("failed to handle message from stream", "message", msg, "error", err)
+			_ = stream.Close()
```
[the remark this answers is not in the document]
Comment (noot) [reply dot/network/service.go:521]: why? we only want to close the stream if there's an error

File: dot/network/connmgr.go [hunk 2]
```diff
@@ -191,9 +196,15 @@ func (cm *ConnManager) Disconnected(n network.Network, c network.Conn) {
 		Addrs: addrs,
 	}
 
-	err := cm.host.connect(info)
-	if err != nil {
-		logger.Warn("failed to reconnect to persistent peer", "peer", c.RemotePeer(), "error", err)
+	for i := 0; i < maxRetries; i++ {
```
[the remark this answers is not in the document]
Comment (noot) [reply dot/network/connmgr.go:199]: this function does run in its own goroutine already AFAIK, so it wouldn't block anything

File: dot/network/sync_test.go [hunk 3]
```diff
@@ -500,7 +502,7 @@ func TestSyncQueue_handleBlockDataFailure_MissingParent(t *testing.T) {
 	q.ctx = context.Background()
 
 	data := testBlockResponseMessage().BlockData
-	q.handleBlockDataFailure(0, fmt.Errorf("failed to get parent hash: Key not found"), data)
+	q.handleBlockDataFailure(0, fmt.Errorf("some error: %w", chaindb.ErrKeyNotFound), data)
```
[the remark this answers is not in the document]
Comment (noot) [reply dot/network/sync_test.go:505]: I wanted to test the `errors.Is` functionality, which unwraps the error for you

File: dot/network/connmgr.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (noot) [reply dot/network/connmgr.go:199]: done

**Comment by edwardmack** [comment]
running `go test ./dot/network -short` is failing on my machine, panic after 10 minute timeout.  Any ideas why?:
```
DBUG[04-13|10:50:53] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
DBUG[04-13|10:50:53] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
WARN[04-13|10:50:54] failed to reconnect to persistent peer   pkg=network peer=12D3KooWHHzSeKaY8xuZVzkLbKFfvNgPPeKhFBGrMbNzbm5akpqu error="context canceled" caller=connmgr.go:202
DBUG[04-13|10:51:18] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
DBUG[04-13|10:51:19] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
DBUG[04-13|10:51:19] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
DBUG[04-13|10:51:20] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
DBUG[04-13|10:51:21] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
DBUG[04-13|10:51:21] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
DBUG[04-13|10:51:22] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
DBUG[04-13|10:51:23] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
DBUG[04-13|10:51:23] peer count                               pkg=network num=0 min=5 max=50 caller=service.go:277
panic: test timed out after 10m0s

goroutine 3544 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1609 +0xe5
created by time.goFunc
	/usr/local/go/src/time/sleep.go:167 +0x45

goroutine 1 [chan receive, 9 minutes]:
testing.(*T).Run(0xc001483b00, 0x11bfab6, 0x13, 0x13267b8, 0x4a4201)
	/usr/local/go/src/testing/testing.go:1160 +0x3ad
testing.runTests.func1(0xc0005a4480)
	/usr/local/go/src/testing/testing.go:1430 +0x78
testing.tRunner(0xc0005a4480, 0xc0006c7d08)
	/usr/local/go/src/testing/testing.go:1108 +0xef

```

**Comment by noot** [comment]
@edwardmack can you pull again and retry? I think I just fixed that but let me know if it's still happening

File: dot/network/notifications.go [hunk 4]
```diff
@@ -161,7 +160,6 @@ func (s *Service) createNotificationsMessageHandler(info *notificationsProtocol,
 				if err != nil {
 					logger.Trace("failed to validate handshake", "protocol", info.protocolID, "peer", peer, "error", err)
 					hsData.validated = false
-					_ = stream.Conn().Close()
```
Comment (edwardmack) [inline dot/network/notifications.go:164]: We no longer close connection stream when there is an error, why?  Is connection closing handled somewhere else?  In network/service.go line 500, the stream is closed when there is an error.  I'm just try to better understand how these streams are handled.

**Comment by edwardmack** [comment]
@noot yes, test ./dot/network -short is passing for me now.

File: dot/network/notifications.go (hunk 4 above)
Comment (noot) [reply dot/network/notifications.go:164]: yes, this was duplicate code, if this function returns an error then `readStream` in service.go will close the stream:
```
		err = handler(stream, msg)
		if err != nil {
			logger.Debug("failed to handle message from stream", "message", msg, "error", err)
			_ = stream.Close()
			return
		}
```

File: dot/network/service.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (arijitAD) [reply dot/network/service.go:521]: Makes sense. I assumed we would close the stream after reading once.

**Review by arijitAD — APPROVED** [review]

**Review by edwardmack — APPROVED** [review]

**Pull request merged**